src/main/java/games/strategy/triplea/delegate/Matches.java

```diff
@@ -2408,41 +2405,30 @@ private Matches() {}
   public static Match<UnitType> unitTypeCanBeInBattle(final boolean attack, final boolean isLandBattle,
       final PlayerID player, final int battleRound, final boolean includeAttackersThatCanNotMove,
       final boolean doNotIncludeAa, final boolean doNotIncludeBombardingSeaUnits) {
-    return Match.of(ut -> {
-      // Filter out anything like factories, or units that have no combat ability AND cannot be taken casualty
-      final Match.CompositeBuilder<UnitType> canBeInBattle = Match.newCompositeBuilder();
-
-      final Match.CompositeBuilder<UnitType> supportOrNotInfrastructureOrAa = Match.newCompositeBuilder();
-      supportOrNotInfrastructureOrAa.add(unitTypeIsInfrastructure().invert());
-      supportOrNotInfrastructureOrAa.add(unitTypeIsSupporterOrHasCombatAbility(attack, player));
-      if (!doNotIncludeAa) {
-        supportOrNotInfrastructureOrAa.add(Match.allOf(unitTypeIsAaForCombatOnly(),
-            unitTypeIsAaThatCanFireOnRound(battleRound)));
-      }
-      canBeInBattle.add(supportOrNotInfrastructureOrAa.any());
-
-      if (attack) {
-        if (!includeAttackersThatCanNotMove) {
-          canBeInBattle.add(unitTypeCanNotMoveDuringCombatMove().invert());
-          canBeInBattle.add(unitTypeCanMove(player));
-        }
-        if (isLandBattle) {
-          if (doNotIncludeBombardingSeaUnits) {
-            canBeInBattle.add(unitTypeIsSea().invert());
-          }
-        } else { // is sea battle
-          canBeInBattle.add(unitTypeIsLand().invert());
-        }
-      } else { // defense
-        if (isLandBattle) {
-          canBeInBattle.add(unitTypeIsSea().invert());
-        } else { // is sea battle
-          canBeInBattle.add(unitTypeIsLand().invert());
+
+    // Filter out anything like factories, or units that have no combat ability AND cannot be taken casualty
+    final PredicateBuilder<UnitType> canBeInBattleBuilder = PredicateBuilder.of(unitTypeIsInfrastructure().invert())
+        .or(unitTypeIsSupporterOrHasCombatAbility(attack, player))
+        .orIf(!doNotIncludeAa, Match.allOf(unitTypeIsAaForCombatOnly(), unitTypeIsAaThatCanFireOnRound(battleRound)));
+
+    if (attack) {
+      if (!includeAttackersThatCanNotMove) {
+        canBeInBattleBuilder
+            .and(unitTypeCanNotMoveDuringCombatMove().invert())
+            .and(unitTypeCanMove(player));
+      }
+      if (isLandBattle) {
+        if (doNotIncludeBombardingSeaUnits) {
+          canBeInBattleBuilder.and(unitTypeIsSea().invert());
         }
+      } else { // is sea battle
+        canBeInBattleBuilder.and(unitTypeIsLand().invert());
       }
+    } else { // defense
+      canBeInBattleBuilder.and((isLandBattle ? unitTypeIsSea() : unitTypeIsLand()).invert());
```

Comment:
Missing invert on unitTypeIsSea()?

@ssoloff @RoiEXLab Can you both double check that this method is equivalent after the refactoring as the diff is hard to follow.

Reply:
Same as above.